### PR TITLE
Develop upstream sync 190316

### PR DESCRIPTION
**Pull request by sumesh13**
Initial test

### CODE REVIEW

**Comment by sumesh13** [comment]
Didnt notice that the commit didnt go in my name :(

**Comment by whchung** [comment]
It seems on `rocm-xla` we are still suffering from 2 issues:

1) `ld.lld` is not found. I imagine `/opt/rocm/hcc/bin` might need to be put into search path.
2) scratch memory doesn't seem to be properly allocated:

```
tensorflow.python.framework.errors_impl.FailedPreconditionError: Failed to memcopy into scratch buffer for device 0
	 [[{{node _SOURCE}}]]
	 [[Compiled_g3dbYwsZ7rc/declustered/_1]]
```

On `cuda` path there is only 1 failure. `rocm` path seems fine.

**Comment by sumesh13** [comment]
Yeah ..am adding the path - hardcoded for now,, Will later see how to get the env variable

**Comment by sumesh13** [comment]
Some tests going out of memory on both rocm and cuda path. I put the ones on the cuda path in the whitelist. At least one of them seems to have a related test disabled in upstream due to going out of memory

File: tensorflow/core/kernels/maxpooling_op.cc [hunk 1]
```diff
@@ -41,9 +41,9 @@ limitations under the License.
 #include "tensorflow/core/util/use_cudnn.h"
 
 #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
-#if GOOGLE_CUDA
-#include "cuda/include/cudnn.h"
-#endif
+#ifdef GOOGLE_CUDA
```
Comment (whchung) [inline tensorflow/core/kernels/maxpooling_op.cc]: @sumesh13 use #if

File: tensorflow/core/kernels/maxpooling_op_gpu.h [hunk 2]
```diff
@@ -14,7 +14,7 @@ limitations under the License.
 ==============================================================================*/
 
 #if !GOOGLE_CUDA && !TENSORFLOW_USE_ROCM
-#error This file must only be included when building with Cuda or ROCm support
+#error This file must only be included when building with Cuda or ROCm  support
```
Comment (whchung) [inline tensorflow/core/kernels/maxpooling_op_gpu.h]: why adding this whitespace?

File: tensorflow/core/kernels/maxpooling_op_gpu.cu.cc [hunk 3]
```diff
@@ -491,9 +500,9 @@ bool MaxPoolGradBackwardNoMask<T>::operator()(
         channels, height, width, kernel_h, kernel_w, stride_h, stride_w, pad_t,
         pad_l, top_diff, bottom_diff);
   } else {
-    GPU_LAUNCH_KERNEL(MaxPoolGradBackwardNoMaskNCHW,
-        dim3(config.block_count), dim3(config.thread_per_block), 0, d.stream(),
-        num_kernels, bottom_data, output_data, pooled_height, pooled_width,
+      GPU_LAUNCH_KERNEL(MaxPoolGradBackwardNoMaskNCHW, dim3(config.block_count),
```
Comment (whchung) [inline tensorflow/core/kernels/maxpooling_op_gpu.cu.cc]: indentation doesn't seem right?

**Comment by whchung** [comment]
@sumesh13 per discussion let's focus on getting `cuda` and `rocm` path sorted out first. I think both paths are converging now. Let's try strive to make this PR merged today (3/21/2019). Please address comments I made for stylistic changes.


For `rocm-xla` path I'm trying some changes in `HIP` so errors related to `hipModuleGetGlobal` be taken care of. We'll need to revise CI scripts in another PR.

**Comment by whchung** [comment]
@sumesh13  in order to drive this PR to a closure and have it merged, please do the followings:

1) push remaining local changes yet submitted to this PR
2) **MAKE SURE** TF CNN benchmarks work
3) analyze and have a rough idea about the symptoms of those failed cases on `rocm` path
4) mark the tests in `rocm` path into the whitelist

**Comment by whchung** [comment]
+@parallelo for awareness

@sumesh13 Per discussion please also change `Dockerfile.rocm` to make sure it contains `ROCm-Device-Libs` with your implementation of `shfl` ( https://github.com/RadeonOpenCompute/ROCm-Device-Libs/commit/7aca7eeb441ece8eff839f0fa8a5bb1682b74897 ). You are likely needed to modify the "dropbox" package built by @parallelo before all bits are shipped in ROCm 2.3.

As we discussed earlier, let's focus on non-XLA paths stable first, so it's ok to have this submitted in a separate PR.

**Comment by sumesh13** [comment]
Tested alexnet/resnet/inception/googlenet - they pass. 

Detailed logs 

Alexnet
Generating training model
Initializing graph
Running warm up
Done warm up
Step    Img/sec total_loss
1       images/sec: 1914.4 +/- 0.0 (jitter = 0.0)       nan
10      images/sec: 1934.3 +/- 4.5 (jitter = 8.9)       nan
20      images/sec: 1930.3 +/- 5.2 (jitter = 7.7)       nan
30      images/sec: 1930.0 +/- 4.1 (jitter = 9.7)       nan
40      images/sec: 1930.8 +/- 3.6 (jitter = 7.7)       nan
50      images/sec: 1930.7 +/- 3.2 (jitter = 8.9)       nan
60      images/sec: 1929.2 +/- 3.1 (jitter = 10.6)      nan
70      images/sec: 1929.4 +/- 2.8 (jitter = 10.2)      nan
80      images/sec: 1929.9 +/- 2.5 (jitter = 8.7)       nan
90      images/sec: 1927.3 +/- 2.5 (jitter = 12.7)      nan
100     images/sec: 1926.7 +/- 2.3 (jitter = 12.1)      nan



googlenet 

Initializing graph
Running warm up
Done warm up
Step    Img/sec total_loss
1       images/sec: 489.5 +/- 0.0 (jitter = 0.0)        7.104
10      images/sec: 486.6 +/- 6.2 (jitter = 2.0)        7.061
20      images/sec: 486.0 +/- 4.1 (jitter = 3.4)        7.082
30      images/sec: 487.6 +/- 2.8 (jitter = 3.0)        7.075
40      images/sec: 486.4 +/- 2.8 (jitter = 2.9)        7.102
50      images/sec: 485.2 +/- 2.7 (jitter = 3.3)        7.053
60      images/sec: 485.9 +/- 2.4 (jitter = 3.4)        7.100
70      images/sec: 485.9 +/- 2.2 (jitter = 3.9)        7.052
80      images/sec: 484.6 +/- 2.2 (jitter = 4.1)        7.043
90      images/sec: 483.7 +/- 2.2 (jitter = 3.8)        7.074
100     images/sec: 483.5 +/- 2.1 (jitter = 4.1)        7.058
----------------------------------------------------------------
total images/sec: 482.60



inception 
ariables:   parameter_server
==========
Generating training model
Initializing graph
Running warm up
Done warm up
Step    Img/sec total_loss
1       images/sec: 121.1 +/- 0.0 (jitter = 0.0)        7.322
10      images/sec: 117.5 +/- 1.0 (jitter = 3.8)        7.319
20      images/sec: 116.9 +/- 0.6 (jitter = 2.2)        7.353
30      images/sec: 117.4 +/- 0.5 (jitter = 2.8)        7.296
40      images/sec: 117.1 +/- 0.5 (jitter = 2.2)        7.273
50      images/sec: 117.0 +/- 0.4 (jitter = 1.9)        7.274
60      images/sec: 117.0 +/- 0.4 (jitter = 2.8)        7.337
70      images/sec: 116.7 +/- 0.3 (jitter = 3.0)        7.288
80      images/sec: 116.7 +/- 0.3 (jitter = 3.1)        7.310
90      images/sec: 116.6 +/- 0.3 (jitter = 2.6)        7.335
100     images/sec: 116.5 +/- 0.3 (jitter = 2.5)        7.370
----------------------------------------------------------------
total images/sec: 116.44



Generating training model
Initializing graph
Running warm up
Done warm up
Step    Img/sec total_loss
1       images/sec: 238.0 +/- 0.0 (jitter = 0.0)        8.220
10      images/sec: 254.0 +/- 1.9 (jitter = 0.7)        7.880
20      images/sec: 252.9 +/- 1.7 (jitter = 0.8)        7.910
30      images/sec: 252.4 +/- 1.3 (jitter = 1.3)        7.820
40      images/sec: 252.9 +/- 1.1 (jitter = 1.3)        8.004
50      images/sec: 252.8 +/- 1.0 (jitter = 1.3)        7.769
60      images/sec: 252.4 +/- 0.9 (jitter = 1.2)        8.113
70      images/sec: 252.7 +/- 0.8 (jitter = 0.9)        7.819
80      images/sec: 252.8 +/- 0.7 (jitter = 0.9)        7.981
90      images/sec: 252.7 +/- 0.7 (jitter = 1.0)        8.100
100     images/sec: 252.7 +/- 0.6 (jitter = 1.1)        8.039

**Comment by sumesh13** [comment]
Pls let me if this looks good for merge

**Comment by whchung** [comment]
Finally we see 2 check marks on both `cuda` and `rocm` path. I'm taking this PR.

**Comment by whchung** [comment]
For `rocm-xla` there are quite a few failed cases compared with the result on #366 .

```
Executed 457 out of 459 tests: 374 tests pass, [0m[31m[1m2 fail to build[0m and [0m[31m[1m83 fail locally[0m.
```

@sumesh13 I expect this would be the priority for you in the next week.

**Pull request merged**